### PR TITLE
Map WooCommerce states and Payment Request states

### LINKED ISSUE #1399: Add states list to match data received from payment request API

### PR DESCRIPTION
**Pull request by ricardo**
Fixes #1399

#### Changes proposed in this Pull Request

**The problem:** Payment Request API states are different than WooCommerce states for some countries. E.g:
`China: Anhui Sheng` and `Anhui / 安徽` in WC. On top of that, WC states are also translated, making it impossible to match in certain scenarios.

This is a temporary solution for matching Payment Request API states and WooCommerce states until we have this list available in core.

I created a class only with a constant array and didn't add its related methods because this is similar to what we're going to get from core when the list is available in a future release of WC.

Right now, to get the current list of translated states, you can do `WC()->countries->get_states( $country )`. In the future, we should be able to do something like `WC()->countries->get_payment_request_states( $country )` and use that instead of this class in the normalization functions.

This PR also introduces a more descriptive error message in case the state field is expected, but not supported by the Payment Request API during checkout. This can happen for many countries and the results are different depending on which browser you're using the Payment Request API.

There can be some cases where there are shipping zones configured for unsupported countries, and in that case, no shipping methods will be available. - This is part of another fix that will be handled in the shipping settings screen, in a different PR.

#### Testing instructions

- Make sure you have the `_wcpay_feature_payment_request` option set to `yes`, and "Payment Request Button" enabled in Payments > Settings.
- Make sure test mode is enabled, so you can make test transactions.
- Go to an HTTPS product URL and click the "Buy now" button.
- Add a shipping or billing address in China, select any state and click "Pay" - notice the order goes through with the correct shipping state
- You can test with other previously unsupported countries: `EG` (Egypt), `HK` (Hong Kong), `IR` (Iran).
- Safari doesn't provide a dropdown list of states, so if you try typing the state code or state name without accents, it should work where it would previously fail. E.g: Country: `BR` (Brazil), State: `Parana` (correct spelling is Paraná). Or `ES` (Spain), State: `Malaga` (correct spelling is Málaga).
- You can also test the error notice for unsupported countries trying to checkout with a shipping address in `AO` (Angola), `BD` (Bangladesh), or `ZA` (South Africa).

-------------------

- [ ] Added changelog entry (or does not apply)
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

### CODE REVIEW

File: includes/constants/class-payment-request-button-states.php [hunk 1]
```diff
@@ -0,0 +1,1162 @@
...
+			'CN24' => [ 'Sichuan Sheng', 'Sichuan Sheng', '四川省' ],
+			'CN25' => [ 'Guizhou Sheng', 'Guizhou Sheng', '贵州省' ],
+			'CN26' => [ 'Shaanxi Sheng', 'Shaanxi Sheng', '陕西省' ],
+			'CN27' => [ 'Gansu Sheng', 'Gansu Sheng', '甘肃省' ],
+			'CN28' => [ 'Qinghai Sheng', 'Qinghai Sheng', '青海省' ],
+			'CN29' => [ 'Ningxia Huizuzizhiqu', 'Ningxia Huizuzizhiqu', '宁夏' ],
+			'CN30' => [ 'Macau', 'Macau', '澳门' ],
+			'CN31' => [ 'Xizang Zizhiqu', 'Xizang Zizhiqu', '西藏' ],
+			'CN32' => [ 'Xinjiang Weiwuerzizhiqu', 'Xinjiang Weiwuerzizhiqu', '新疆' ],
+			// [ 'Taiwan', 'Taiwan', '台湾' ],
```
Comment (ricardo) [inline includes/constants/class-payment-request-button-states.php:192]: States that do not match with any WC state, or that are remaining, are commented out.

**Comment by mattallan** [comment]
Nice one @ricardo !! Code LGTM :+1:

I ran these changes through the checkout using the following data:
 - [x] Germany (no states)
   - Street: Bleibtreustraße 45
   - City: Schwanau
   - State: Baden-Württemberg
   - Postcode: 77963
 - [x] China (3 state options and some with non-ascii characters)
   - Street: Long Gang You Er Yuan
   - District: Hongqiao
   - City: City Area
   - Province: Tianjin Shi
   - Postcode: 300301
   - Phone: 13092632818
 - [x] Australia (2 states options)
 - [x] US Armed Forces (AP)
   - Street: USCGC HAMILTON
   - City: FPO
   - State: Armed Forces (AP)
   - Postcode: 96667
 - [x] United States (default)
   - Street: 959 Market St
   - City: San Francisco
   - State: California
   - Postcode: 94103
 - [x] South Africa ([Confirmed error message](https://d.pr/i/I69nhd))
   - Street: 1874 North Street
   - City: Zeerust
   - Postcode: 2876
   - Phone: 083 334 1031
 - [ ] Japan :x:
   - Street: 139-1126, Senju Asahicho
   - Prefecture: Tokyo
   - Postcode: 120-0015
   - Phone: +8131-461-6948

Before testing China, I tried testing with Japan but I kept running into the following checkout error: https://d.pr/i/DlMDkb. Looking more into this, it looks like [WC asks for a city on the checkout](https://d.pr/i/JswaLz), but the [Google Pay window doesn't ask for this.](https://d.pr/i/JQ8URJ)

Is this a known issue and should it be fixed in this PR or another issue? Seems unrelated to state mappings so I'm happy for this to be merged as-is.

Nice work on this one 💯

**Review by mattallan — APPROVED** [review]
Nice one @ricardo !! Code LGTM :+1:

I ran these changes through the checkout using the following data:
 - [x] Germany (no states)
   - Street: Bleibtreustraße 45
   - City: Schwanau
   - State: Baden-Württemberg
   - Postcode: 77963
 - [x] China (3 state options and some with non-ascii characters)
   - Street: Long Gang You Er Yuan
   - District: Hongqiao
   - City: City Area
   - Province: Tianjin Shi
   - Postcode: 300301
   - Phone: 13092632818
 - [x] Australia (2 states options)
 - [x] US Armed Forces (AP)
   - Street: USCGC HAMILTON
   - City: FPO
   - State: Armed Forces (AP)
   - Postcode: 96667
 - [x] United States (default)
   - Street: 959 Market St
   - City: San Francisco
   - State: California
   - Postcode: 94103
 - [x] South Africa ([Confirmed error message](https://d.pr/i/I69nhd))
   - Street: 1874 North Street
   - City: Zeerust
   - Postcode: 2876
   - Phone: 083 334 1031
 - [ ] Japan :x:
   - Street: 139-1126, Senju Asahicho
   - Prefecture: Tokyo
   - Postcode: 120-0015
   - Phone: +8131-461-6948

Before testing China, I tried testing with Japan but I kept running into the following checkout error: https://d.pr/i/DlMDkb. Looking more into this, it looks like [WC asks for a city on the checkout](https://d.pr/i/JswaLz), but the [Google Pay window doesn't ask for this.](https://d.pr/i/JQ8URJ)

Is this a known issue and should it be fixed in this PR or another issue? Seems unrelated to state mappings so I'm happy for this to be merged as-is.

Nice work on this one 💯

**Comment by ricardo** [comment]
@mattallan thank you so much for the review and the thorough testing! 🙇‍♂️ 

> Before testing China, I tried testing with Japan but I kept running into the following checkout error: https://d.pr/i/DlMDkb. Looking more into this, it looks like WC asks for a city on the checkout, but the Google Pay window doesn't ask for this.

Yeah, Japan + Chrome is a known issue and a very distinct case. There are only 4 countries where the city field is missing, but required in WC. I added my discoveries in [this comment](https://github.com/Automattic/woocommerce-payments/blob/067d8281c96315a65a749d8f8ffa56cc3a39dfd3/includes/constants/class-payment-request-button-states.php#L31). I've previously opened an issue in WC core to make the city field not required in Japan ([#29373](https://github.com/woocommerce/woocommerce/issues/29373)), but I'll add the other 3 countries to the issue.

I wish there was a better way to handle the nuances in validation between WC and the Payment Request API, but one day we will get there 🙏

**Comment by mattallan** [comment]
> I added my discoveries in this comment

Aha! Sorry I missed this comment in the code!

Given that, this is good to be merged! Nice work again on this one @ricardo 💃

**Comment by ricardo** [comment]
@mattallan Sorry to push a commit after you've approved the PR. I just noticed a minor change I made to the previous implementation in `get_normalized_state_from_wc_states` where I did the `preg_match` from the input, instead of keeping the previous approach. I just tested the critical flows again and nothing has changed. Just a minor tweak 🤏

**Pull request merged**